ICE: `in setup_associated_types, at rust/typecheck/rust-hir-trait-resolve.cc:654`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code
Reduced
```rust
trait Trait<T: NewTrait> {}
trait NewTrait {}
fn foo<T>()
where
    T: Trait< fn()>,
{}
impl<'a> Trait<fn()> for () {}
fn main() {
    foo::<()>();
}
```
Original
```rust
trait Trait<T: NewTrait> {}
fn foo<T>()
where
    T: Trait<for<'b> fn(&'b u32)>,
{}
impl<'a> Trait<fn(&'a u32)> for () {}
fn main() {
    foo::<()>();
}
trait NewTrait {
    type Assoc;
    const C: Self::Assoc;
    fn f(&self) -> Self::Assoc;
}

```



### Meta

1a91028

I've checked #3931; it is no longer an ICE
### Error output

```
/tmp/tmp.dtqrJ08387/nocore_wrapper.rs:13:16: error: bounds not satisfied for abi:rust fnptr () -> () 'NewTrait' is not satisfied [E0277]
    7 | trait Trait<T: NewTrait> {}
      |                ~~~~~~~~
......
   13 | impl<'a> Trait<fn()> for () {}
      |                ^~
/tmp/tmp.dtqrJ08387/nocore_wrapper.rs:11:15: error: bounds not satisfied for abi:rust fnptr () -> () 'NewTrait' is not satisfied [E0277]
    7 | trait Trait<T: NewTrait> {}
      |                ~~~~~~~~
......
   11 |     T: Trait< fn()>,
      |               ^~
/tmp/tmp.dtqrJ08387/nocore_wrapper.rs:13:10: error: bounds not satisfied for () 'NewTrait' is not satisfied [E0277]
    7 | trait Trait<T: NewTrait> {}
      |                ~~~~~~~~
......
   13 | impl<'a> Trait<fn()> for () {}
      |          ^~~~~
```

<!--
Include a backtrace if available.
-->
<details><summary>Backtrace</summary>
  <p>
  
  ```
crab1: internal compiler error: in setup_associated_types, at rust/typecheck/rust-hir-trait-resolve.cc:654
0x2c1ef7b internal_error(char const*, ...)
        ../../gcc/diagnostic-global-context.cc:787
0xb49f7d fancy_abort(char const*, int, char const*)
        ../../gcc/diagnostics/context.cc:1813
0x910019 Rust::Resolver::AssociatedImplTrait::setup_associated_types(Rust::TyTy::BaseType const*, Rust::TyTy::TypeBoundPredicate const&, Rust::TyTy::SubstitutionArgumentMappings*, bool)
        ../../gcc/rust/typecheck/rust-hir-trait-resolve.cc:654
0xe6fee7 Rust::TyTy::SubstitutionRef::monomorphize()
        ../../gcc/rust/typecheck/rust-tyty-subst.cc:1103
0xe6b48e Rust::TyTy::TypeCheckCallExpr::visit(Rust::TyTy::FnType&)
        ../../gcc/rust/typecheck/rust-tyty-call.cc:253
0xeb06ef Rust::TyTy::TypeCheckCallExpr::go(Rust::TyTy::BaseType*, Rust::HIR::CallExpr&, Rust::TyTy::VariantDef&, Rust::Resolver::TypeCheckContext*)
        ../../gcc/rust/typecheck/rust-tyty-call.h:39
0xeb06ef Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::CallExpr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:274
0xeabe1c Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:51
0xeb8910 ???
        ../../gcc/rust/typecheck/rust-hir-type-check-stmt.h:33
0xeb910b Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt&)
        ../../gcc/rust/typecheck/rust-hir-type-check-stmt.cc:34
0xead5a1 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:638
0xeabe1c Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
        ../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:51
0xe9827b Rust::Resolver::TypeCheckItem::visit(Rust::HIR::Function&)
        ../../gcc/rust/typecheck/rust-hir-type-check-item.cc:598
0xe9190d Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
        ../../gcc/rust/typecheck/rust-hir-type-check-item.cc:57
0xe9190d Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
        ../../gcc/rust/typecheck/rust-hir-type-check-item.cc:43
0xe4c12b Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
        ../../gcc/rust/typecheck/rust-hir-type-check.cc:71
0xd11bcf Rust::Session::compile_crate(char const*)
        ../../gcc/rust/rust-session-manager.cc:763
/home/gccrs/build/gcc/crab1 /tmp/tmp.dtqrJ08387/nocore_wrapper.rs -frust-incomplete-and-experimental-compiler-do-not-use -c
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
  ```
  
  </p>
</details>

--- a/gcc/testsuite/rust/compile/issue-4481.rs
+++ b/gcc/testsuite/rust/compile/issue-4481.rs
@@ -1,0 +1,25 @@
+#![feature(no_core, intrinsics, staged_api, lang_items)]
+#![no_core]
+
+#[lang = "sized"]
+pub trait Sized {}
+
+trait Trait<T: NewTrait> {}
+
+fn foo<T>()
+where
+    T: Trait<for<'b> fn(&'b u32)>, // { dg-error {bounds not satisfied for abi:rust fnptr .& u32 ,. -> .. .NewTrait. is not satisfied \[E0277\]} }
+{
+}
+
+impl<'a> Trait<fn(&'a u32)> for () {} // { dg-error {bounds not satisfied for abi:rust fnptr .& u32 ,. -> .. .NewTrait. is not satisfied \[E0277\]} }
+
+fn main() {
+    foo::<()>();
+}
+
+trait NewTrait {
+    type Assoc;
+    const C: Self::Assoc;
+    fn f(&self) -> Self::Assoc;
+}

